File: kotlin/src/main/kotlin/aether/routing/MeshSender.kt
// SPDX-License-Identifier: MIT

package aether.routing

import aether.models.PeerInfo
import aether.protocol.MeshPacket

/**
 * Minimal sending abstraction for routing/DTN/SOS. Hosts wire this with a thin
 * adapter over their transport so the protocol services do not take a hard
 * dependency on a specific transport implementation.
 */
interface MeshSender {
    /** The local node's UHID. Used as packet.sourceUhid on outbound packets. */
    val localUhid: String

    /** Local node's last-known geohash, or null if not shared. */
    val localGeohash: String? get() = null

    /** Snapshot of currently directly-connected peers. */
    fun connectedPeers(): List<PeerInfo> = emptyList()

    /** Forward a packet to a single next-hop peer. Returns true if delivered. */
    suspend fun send(packet: MeshPacket, nextHopUhid: String): Boolean

    /** Broadcast a packet to every connected peer. Returns the fan-out count. */
    suspend fun broadcast(packet: MeshPacket): Int
}

File: swift/Sources/AetherProtocol/Routing/MeshSender.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Minimal sending abstraction for routing/DTN/SOS. Hosts wire this with a thin
/// adapter over their transport so the protocol services do not take a hard
/// dependency on a specific transport implementation.
public protocol MeshSender: Sendable {
    /// The local node's UHID. Used as packet.sourceUhid on outbound packets.
    var localUhid: String { get }

    /// Local node's last-known geohash, or nil if not shared.
    var localGeohash: String? { get }

    /// Snapshot of currently directly-connected peers.
    func connectedPeers() -> [PeerInfo]

    /// Forward a packet to a single next-hop peer.
    func send(_ packet: MeshPacket, nextHopUhid: String) async -> Bool

    /// Broadcast a packet to every connected peer; returns the fan-out count.
    func broadcast(_ packet: MeshPacket) async -> Int
}

public extension MeshSender {
    var localGeohash: String? { nil }
    func connectedPeers() -> [PeerInfo] { [] }
}
